ec/ecresource/organizationresource/resource_test.go

```diff
@@ -0,0 +1,646 @@
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.email", "newuser@example.com"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.invitation_pending", "false"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.organization_role", "organization-admin"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.user_id", "userid2"),
+				),
+			},
+			// Adding roles to member
+			{
+				Config: configWithAddedRoles,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "members.%", "2"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.email", "newuser@example.com"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.organization_role", "organization-admin"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.deployment_roles.0.role", "editor"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.deployment_roles.0.deployment_ids.0", "abc"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.deployment_roles.1.role", "viewer"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.deployment_roles.1.all_deployments", "true"),
+				),
+			},
+			// Removing roles from member
+			{
+				Config: configWithRemovedRoles,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "members.%", "2"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.email", "newuser@example.com"),
+					resource.TestCheckNoResourceAttr(resourceName, "members.newuser@example.com.organization_role"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.deployment_roles.0.role", "viewer"),
+					resource.TestCheckResourceAttr(resourceName, "members.newuser@example.com.deployment_roles.0.all_deployments", "true"),
+				),
+			},
+			// Removing member from organization
+			{
+				Config: baseConfig,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "members.%", "1"),
+					resource.TestCheckNoResourceAttr(resourceName, "members.newuser@example.com"),
+				),
+			},
+			// Invite member
+			{
+				Config: configWithNewMember,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "members.%", "2"),
+				),
+			},
+			// Un-invite member (where the member is removed before they have accepted the invitation)
+			{
+				Config: baseConfig,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "members.%", "1"),
+					resource.TestCheckNoResourceAttr(resourceName, "members.newuser@example.com"),
+				),
+			},
+		},
+	})
+}
+
+func mockApi() *api.API {
```

Comment:
I think the main mock function is quite important for the test: The list of mocked responses is chosen to match the list of testcases, so every testcase has their exact mocks.

However I could see moving mocking to another file being better for readability, since it makes it easier to have the mocks open in a separate view.

Reply:
I'm now also thinking if I should split the tests into separate cases. Since they are all mocked anyway, there is no need to have one huge testcase.

Reply:
I split up all the cases, this will also make it much easier to investigate what is actually failing (though it does add some boilerplate)